nimble/host/src/ble_gap.c

```diff
@@ -1295,106 +1305,134 @@ ble_gap_rx_adv_set_terminated(struct hci_le_adv_set_terminated *evt)
-    event.periodic_adv_sync_estab.sid                = evt->sid;
-    event.periodic_adv_sync_estab.adv_addr_type      = evt->adv_addr_type;
+        event.periodic_sync.sync_handle = evt->sync_handle;
+        event.periodic_sync.sid = evt->sid;
+        event.periodic_sync.adv_addr = ble_gap_sync.psync->advertiser_addr;
+        event.periodic_sync.adv_phy = evt->adv_phy;
+        event.periodic_sync.per_adv_ival = evt->per_adv_ival;
+        event.periodic_sync.adv_clk_accuracy = evt->adv_clk_accuracy;
 
-    memcpy(event.periodic_adv_sync_estab.adv_addr, evt->adv_addr, 6);
+        ble_hs_periodic_sync_insert(ble_gap_sync.psync);
+    } else {
+        ble_hs_periodic_sync_free(ble_gap_sync.psync);
+    }
 
-    event.periodic_adv_sync_estab.adv_phy            = evt->adv_phy;
-    event.periodic_adv_sync_estab.per_adv_ival       = evt->per_adv_ival;
-    event.periodic_adv_sync_estab.adv_clk_accuracy   = evt->adv_clk_accuracy;
+    cb = ble_gap_sync.cb;
+    cb_arg = ble_gap_sync.cb_arg;
 
-    ble_gap_master.pending_create_sync = 0;
+    ble_gap_sync.op = BLE_GAP_OP_NULL;
+    ble_gap_sync.cb_arg = NULL;
+    ble_gap_sync.cb_arg = NULL;
+    ble_gap_sync.psync = NULL;
 
-    ble_gap_master_extract_state(&state, 0);
-    if (ble_gap_has_client(&state)) {
-        state.cb(&event, state.cb_arg);
+    ble_hs_unlock();
+
+    ble_gap_event_listener_call(&event);
+    if (cb) {
+        cb(&event, cb_arg);
     }
 }
 
 void
 ble_gap_rx_periodic_adv_rpt(struct hci_le_subev_periodic_adv_rpt *evt)
 {
+    struct ble_hs_periodic_sync *psync;
     struct ble_gap_event event;
-    struct ble_gap_master_state state;
+    ble_gap_event_fn *cb;
+    void *cb_arg;
+
+    ble_hs_lock();
+    psync = ble_hs_periodic_sync_find_by_handle(evt->sync_handle);
+    if (psync) {
+        cb = psync->cb;
+        cb_arg = psync->cb_arg;
+    }
+    ble_hs_unlock();
+
+    if (!psync) {
+        return;
+    }
 
     memset(&event, 0, sizeof event);
 
-    event.type = BLE_GAP_EVENT_PERIODIC_DISC;
-    event.periodic_disc.sync_handle = evt->sync_handle;
-    event.periodic_disc.tx_power = evt->tx_power;
-    event.periodic_disc.rssi = evt->rssi;
-    event.periodic_disc.data_status = evt->data_status;
-    event.periodic_disc.data_length = evt->data_length;
-    event.periodic_disc.data = evt->data;
+    event.type = BLE_GAP_EVENT_PERIODIC_REPORT;
+    event.periodic_report.sync_handle = evt->sync_handle;
+    event.periodic_report.tx_power = evt->tx_power;
+    event.periodic_report.rssi = evt->rssi;
+    event.periodic_report.data_status = evt->data_status;
+    event.periodic_report.data_length = evt->data_length;
+    event.periodic_report.data = evt->data;
 
-    ble_gap_master_extract_state(&state, 0);
-    if (ble_gap_has_client(&state)) {
-        state.cb(&event, state.cb_arg);
+    /* TODO should we allow for listener too? this can be spammy and is more
+     * like ACL data, not general event
+     */
+    if (cb) {
```

Comment:
you can put this in line 1379 since there's no point in creating event which is not used later